### PR TITLE
Add blog post for Pandas support in PyFlink.

### PR DESCRIPTION
**Pull request by sunjincheng121**
Flink community put some great effort in integrating Pandas into PyFlink with the latest Flink version 1.11. Some of the added features include support for Pandas UDF and the conversion between Pandas DataFrame and Table. In this blog, we will introduce how these functionalities work and how to use them with a step-by-step example.

### CODE REVIEW

**Comment by sunjincheng121** [comment]
@MarkSfik @dianfu Would be great if you have time to have look at the PR :).

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (dianfu) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
Python has evolved into one of the most important programming languages for many fields of data processing. So big has been Python’s popularity, that it has pretty much become the default data processing language for data scientists. On top of that, there is a plethora of Python-based data processing tools such as NumPy, Pandas, and Scikit-learn that have gained additional popularity due to their flexibility or powerful functionalities. 
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (dianfu) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]: follows -> the following?

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (dianfu) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]: It seems that the double quotation is in Chinese, should use English double quotation?

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (dianfu) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
To submit the job:
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (dianfu) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
The community will focus on adding more features and bringing additional optimizations in the next releases. Such optimizations and additions include Python DataStream API and more integration with the Python ecosystem, such as support for distributed Pandas in Flink. Stay tuned for more information and updates with the upcoming releases!
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (dianfu) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md:25]: remainder -> remaining?

**Review by dianfu — COMMENTED** [review]
@sunjincheng121 Thanks a lot for the great work! It looks good to be overall with just a few minor comments.

**Comment by sunjincheng121** [comment]
Thanks for the quick review @dianfu 
I have update the PR accordingly. :)

**Comment by dianfu** [comment]
LGTM. +1 from my side.

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
excerpt: The Apache Flink community put some great effort into integrating Pandas with PyFlink in the latest Flink version 1.11. Some of the added features include support for Pandas UDF and the conversion between Pandas DataFrame and Table. In this article, we will introduce how these functionalities work and how to use them with a step-by-step example. 
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
In an effort to meet the user needs and demands, the Flink community hopes to leverage and make better use of these tools.  Along this direction, the Flink community put some great effort in integrating Pandas into PyFlink with the latest Flink version 1.11. Some of the added features include support for Pandas UDF and the conversion between Pandas DataFrame and Table. Pandas UDF not only greatly improve the execution performance of Python UDF, but also make it more convenient for users to leverage libraries such as Pandas and NumPy in Python UDF. Additionally, providing support for the conversion between Pandas DataFrame and Table enables users to switch processing engines seamlessly without the need for an intermediate connector. The following sections will introduce how these functionalities work and how to use them with a step-by-step example.
```

Comment (MarkSfik) [reply _posts/2020-07-28-pyflink-pandas-support-flink.md]: I agree with @dianfu that "remainder" sounds a bit strange here. I made a suggestion for an alternative.

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
The introduction of Pandas UDF is used to address these drawbacks. For Pandas UDF, a batch of rows is transferred between the JVM and PVM in a columnar format ([Arrow memory format](https://arrow.apache.org/docs/format/Columnar.html)). The batch of rows will be converted into a collection of Pandas Series and will be transferred to the Pandas UDF to then leverage popular Python libraries (such as Pandas, or NumPy) for the Python UDF implementation.
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
The performance of vectorized UDFs is usually much higher when compared to the normal Python UDF, as the serialization/deserialization overhead is minimized by falling back to [Apache Arrow](https://arrow.apache.org/), while handling `pandas.Series` as input/output allows us to take full advantage of the Pandas and NumPy libraries, making it a popular solution to parallelize Machine Learning and other large-scale, distributed data science workloads (e.g. feature engineering, distributed model application).
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
Pandas DataFrame is the de-facto standard for working with tabular data in the Python community while PyFlink Table is Flink’s representation of the tabular data in Python. Enabling the conversion between PyFlink Table and Pandas DataFrame allows switching between PyFlink and Pandas seamlessly when processing data in Python. Users can process data by utilizing one execution engine and switch to a different one effortlessly. For example, in case users already have a Pandas DataFrame at hand and want to perform some expensive transformation, they can easily convert it to a PyFlink Table and leverage the power of the Flink engine. On the other hand, users can also convert a PyFlink Table to a Pandas DataFrame and perform the same transformation with the rich functionalities provided by the Pandas ecosystem.
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
If you don't have a version above 3.5, you can use virtualenv with the following commands:
```

Comment (MarkSfik) [reply _posts/2020-07-28-pyflink-pandas-support-flink.md]: I agree @dianfu. I made a suggestion to address this. Thanks!

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
Pandas UDFs take `pandas.Series` as the input and return a `pandas.Series` of the same length as the output. Pandas UDFs can be used at the exact same place where non-Pandas functions are currently being utilized. To mark a UDF as a Pandas UDF, you only need to add an extra parameter udf_type=”pandas” in the udf decorator:
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
The Pandas UDF above uses the Pandas `dataframe.interpolate()` function to interpolate the missing temperature data for each equipment id. This is a common IoT scenario whereby each equipment/device reports it’s id and temperature to be analyzed, but the temperature field may be null due to various reasons.
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
- Finally, you can see the execution result on the command line. As you can see, all the temperature data with an empty value has been interpolated:
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
You can use the `from_pandas()` method to create a PyFlink Table from a Pandas DataFrame or use the `to_pandas()` method to convert a PyFlink Table to a Pandas DataFrame.
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (MarkSfik) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
In this article, we introduce the integration of Pandas in Flink 1.11, including Pandas UDF and the conversion between Table and Pandas. In fact, in the latest Apache Flink release, there are many excellent features added to PyFlink, such as support of User-defined Table functions and User-defined Metrics for Python UDFs. What’s more, from Flink 1.11, you can build PyFlink with Cython support and “Cythonize” your Python UDFs to substantially improve code execution speed (up to 30x faster, compared to Python UDFs in Flink 1.10).
```

**Review by MarkSfik — COMMENTED** [review]
Thanks for the comments @dianfu and for adding this @sunjincheng121 
I left some minor suggestions for a final check and review.

**Comment by sunjincheng121** [comment]
Thanks for the nice comments @MarkSfik

**Comment by morsapaes** [comment]
Thanks for writing this blogpost, @sunjincheng121. I didn't reproduce the PR locally to have a closer look at the content, but the code seems to be using some deprecated methods of the Table API (e.g. insert_into, execute) — could you update it to use the new interfaces?

**Comment by sunjincheng121** [comment]
> Thanks for writing this blogpost, @sunjincheng121. I didn't reproduce the PR locally to have a closer look at the content, but the code seems to be using some deprecated methods of the Table API (e.g. insert_into, execute) — could you update it to use the new interfaces?

Hi @morsapaes, Thank you for your reminding. Because the new API is asynchronous submission, So, in the current example, we cannot get the result in python shell, to keep the examples simple, I prefer to use `insert_into` and `tenv.execute()` in this blog. :)

**Comment by morsapaes** [comment]
Aaaaah! I was trying it out in PyCharm! 🐍  Maybe we could add an annotation that makes that clear, then? In case someone also does it elsewhere.

**Comment by sunjincheng121** [comment]
> Aaaaah! I was trying it out in PyCharm! 🐍 Maybe we could add an annotation that makes that clear, then? In case someone also does it elsewhere.

We can add more annotation in the doc, and will do that in : https://cwiki.apache.org/confluence/display/FLINK/FLIP-133%3A+Rework+PyFlink+Documentation 
What do you think?:)

**Comment by dianfu** [comment]
Thanks @sunjincheng121 a lot for the PR and @MarkSfik @morsapaes for the great review. Regarding to the deprecated API used in the example, I agree with @sunjincheng121 and also prefer to use the deprecated API here and add the detailed documentation about the new API in the PyFlink doc as the behavior of the new API is a little different between executing a job in MiniCluster(e.g. IDE) and submitting to a remote cluster. We need add more documentation about the difference.

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
<img src="{{ site.baseurl }}/img/blog/2020-07-28-pyflink-pandas/python-scientific-stack.png" width="450px" alt="Python Scientific Stack"/>
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
In an effort to meet the user needs and demands, the Flink community hopes to leverage and make better use of these tools.  Along this direction, the Flink community put some great effort in integrating Pandas into PyFlink with the latest Flink version 1.11. Some of the added features include **support for Pandas UDF** and the **conversion between Pandas DataFrame and Table**. Pandas UDF not only greatly improve the execution performance of Python UDF, but also make it more convenient for users to leverage libraries such as Pandas and NumPy in Python UDF. Additionally, providing support for the conversion between Pandas DataFrame and Table enables users to switch processing engines seamlessly without the need for an intermediate connector. In the remainder of this article, we will introduce how these functionalities work and how to use them with a step-by-step example.
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
<img src="{{ site.baseurl }}/img/blog/2020-07-28-pyflink-pandas/vm-communication.png" width="550px" alt="VM Communication"/>
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
Using Python in Apache Flink requires installing PyFlink, which is available on [PyPI](https://pypi.org/project/apache-flink/) and can be easily installed using `pip`. Before installing PyFlink, check the working version of Python running in your system using:
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
Please note that Python 3.5 or higher is required to install and run PyFlink.
</div>
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
# Examples

### Installation
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion

```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [reply _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
If you don't have a version above 3.5, you can create a virtual environment (i.e. [virtualenv](https://docs.python-guide.org/dev/virtualenvs/)) with the following commands:
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion

And then install PyFlink:
```

File: _posts/2020-07-28-pyflink-pandas-support-flink.md
Comment (morsapaes) [inline _posts/2020-07-28-pyflink-pandas-support-flink.md]:
```suggestion
```

**Review by morsapaes — COMMENTED** [review]
Thanks for the details, @dianfu and @sunjincheng121. IMO, the instinct of a user is to try these things out in the IDE and not on the console, so this can be confusing if there's no documentation, indeed.

Added a few comments after building the website. Also, it seems like there is an extra file committed (_posts/2020-07-28-pyflink-pandas-udf-support-flink.md) that doesn't need to be there. 😉

**Comment by sunjincheng121** [comment]
Thanks for the review @morsapaes ,I have update the PR :)

**Comment by morsapaes** [comment]
Thanks, Jincheng! @dianfu, can you +1 the final version and merge maybe tomorrow (Tuesday is usually a better day to post)?

**Review by dianfu — APPROVED** [review]
@morsapaes Thanks for the review and thanks @sunjincheng121 for this PR. The latest version looks really good to me. +1 from my side. I'll merge it tomorrow if there are no more feedback.

**Comment by sunjincheng121** [comment]
Thanks for the detail review @morsapaes @dianfu . 
Hi Dian, before merging we can have a little bit discussion about  virtual env part,  because，if `python` command can't launch the python 3 interpreter, we have many ways to solve it. We can leave space for users or have more detail info in the PR.

**Comment by dianfu** [comment]
@sunjincheng121 As discussed offline, I'll remove the part about how to setup Python virtual environment during merging.

**Comment by morsapaes** [comment]
Thanks, @dianfu ! 🎈  Just a reminder for the future to update the blogpost date (and, if possible, also the name of the file/images).

**Comment by dianfu** [comment]
@morsapaes Thanks a lot for the kind reminder. Updated the blogpost date.

**Pull request closed without merging**
